### PR TITLE
Buckets

### LINKED ISSUE #11125: WTB: Carryable non-backpackable items, aka buckets
Buckets are a fairly common container for various liquids in real life, but they make no appearances in Cataclysm due to being a fairly useless thing to have -- being non-sealable means they either couldn't be carryable when full, or would allow you to have literal buckets of water in your backpack. A related issue already occurs with things like lit charcoal furnaces, which you _can_ carry around in your backpack despite being presumably quite hot and possibly dangerous.

A proposed solution would be to have one or two flags that would allow us to say that _an item may be carried in one's hands, **but not in one's main inventory**_, optionally only when the item has liquid contents. This would also allow us to use tin cans and aluminum cans as impromptu water buckets, because who doesn't want to carry their water to the fire for boiling one charge at a time?

Of special note: cardboard boxes probably shouldn't be able to carry water. Just saying.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/8286179-wtb-carryable-non-backpackable-items-aka-buckets?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Coolthulhu**
Bucket is a container that you can't place in the inventory or on a vehicle without emptying it first.

For example a pot: you can carry it, you can fill it with water, you can carry it in your hands while it's filled with water, but putting it in your backpack while it's filled with water is a bad idea.

A bucket is any watertight container that is not sealable and that does not preserve contents. The latter condition is needed because at the moment all non-empty cans are sealed (even if you ate half of its contents), so if cans were buckets, you could refill them with meat soup and have them preserve the newly poured soup without using a can sealer.

EDIT: Doesn't handle advanced inventory too well yet (makes it spill stuff without a warning). Should that be in before it is mergeable?

EDIT 2: Closes #11125

### CODE REVIEW

**Comment by Rivet-the-Zombie** [comment]
This is a great idea!

**Comment by narc0tiq** [comment]
You can probably put "closes  #11125" in the OP.

**Comment by Mshock777** [comment]
Will it be possible to make bowls and some other stuff from containers work like this in the future?

**Comment by Coolthulhu** [comment]
Yes.

Any non-sealable container that doesn't preserve contents from rot (cans do it) is automatically a bucket. Non-watertight buckets are useless at the moment (powders don't need a container), but can exist.

File: src/game.cpp [hunk 1]
```diff
@@ -10227,9 +10227,11 @@ bool game::handle_liquid(item &liquid, bool from_ground, bool infinite, item *so
         }
         return true;
 
-    } else { // filling up normal containers
+    } else {
+        // Filling up normal containers
+        bool allow_bucket = cont == &g->u.weapon || !g->u.has_item( *cont );
```
Comment (codemime) [inline src/game.cpp]: No need to mention `g->` since we're already inside `game` (here and below).

**Comment by SeanMirrsen** [comment]
I like the concept. How open would it be to further extension? What I mean is, containers have a lot of variables. Does it let water through (fit for carrying liquids/watertight)? Does it let air through (fit for preserving food/airtight)? Is it insulated (maintains temperature)? Can it be carried in a backpack (sealable)? Does it expand as it is filled? Etc, etc. A system of variables.

**Comment by Coolthulhu** [comment]
There could be more restrictive buckets that also spill their stuff when placed on the ground. For example, plastic bags when used to carry liquids - you can hold water in a bag, but it has to hang.

More items may become buckets later on. Cans (when split into closed and open cans), bowls, glasses. New buckets may be added (for example, actual buckets).

I added stubs for content restriction for containers, but I don't have a good idea on what to do with that. Other than making concentrated acid require glass containers.

**Comment by Mshock777** [comment]
Also, would it be possible and reasonable to make these items collect water during rain?

File: src/character.cpp [hunk 2]
```diff
@@ -566,14 +566,17 @@ std::vector<const item *> Character::items_with( const std::function<bool(const
...
 {
- itype_id item_type_id = "null";
- if( it.type ) item_type_id = it.type->id;
+    itype_id item_type_id = "null";
+    if( it.type ) {
+        item_type_id = it.type->id;
+    }
 
- last_item = item_type_id;
+    last_item = item_type_id;
```
Comment (BevapDin) [inline src/character.cpp]: You can get rid of that local variable and assign `it.typeId()` directly. `item::type` should never be null, and even if, `typeId()` would take care of that.

File: src/item.cpp [hunk 3]
```diff
@@ -3542,6 +3565,60 @@ bool item::is_artifact() const
...
+    }
+
+    // @todo Acid in waterskins
+    return true;
+}
+
+bool item::spill_contents( Character &c )
+{
+    if( c.is_npc() ) {
+        spill_contents( c.pos() );
```
Comment (BevapDin) [inline src/item.cpp]: The return value of that function call is not used, is that intended? Should this return right here and skip the remaining parts of the function?

File: src/pickup.cpp [hunk 4]
```diff
@@ -271,12 +272,58 @@ static bool select_autopickup_items( std::vector<item> &here, std::vector<pickup
...
+
+    amenu.selected = 0;
+    amenu.text = explain;
+
+    offered_swap = true;
+    // @todo Gray out if not enough hands
+    amenu.addentry( WIELD, !g->u.weapon.has_flag( "NO_UNWIELD" ), 'w',
+                    _("Dispose of %s and wield %s"), g->u.weapon.display_name().c_str(),
+                    it.display_name().c_str() );
+    if( it.is_armor() ) {
```
Comment (BevapDin) [inline src/pickup.cpp]: There's a `player::can_wear` function that may be useful here.

File: src/pickup.cpp [hunk 5]
```diff
@@ -271,12 +272,58 @@ static bool select_autopickup_items( std::vector<item> &here, std::vector<pickup
...
+    WIELD,
+    WEAR,
+    SPILL,
+    STASH,
+    NUM_ANSWERS
+};
+
+pickup_answer handle_problematic_pickup( const item &it, bool &offered_swap, const std::string &explain )
+{
+    if( !g->u.is_armed() && !g->u.keep_hands_free ) {
```
Comment (BevapDin) [inline src/pickup.cpp]: Extracting `g->u` into a local variable `player &u = g->u;` makes the code a bit easier to read. But more importantly, it allows to make that reference a parameter when needed later.

File: src/advanced_inv.cpp [hunk 6]
```diff
@@ -1985,7 +1985,8 @@ bool advanced_inventory::move_content( item &src_container, item &dest_container
     }
 
     std::string err;
-    if( !dest_container.fill_with( src, err ) ) {
+    // @todo Allow buckets here, but require them to be on the ground or wielded
+    if( !dest_container.fill_with( src, err, false ) ) {
```
Comment (BevapDin) [inline src/advanced_inv.cpp]: This parameter being a boolean is the only thing I don't like about the implementation. I just can't tell the difference between `fill_with( src, err, true )` and `fill_with( src, err, false )`. At least not without looking up the function definition.

Comment (Coolthulhu) [inline src/advanced_inv.cpp]: How should it be done? Should I add an enum describing the requirements or something like that?

Comment (BevapDin) [inline src/advanced_inv.cpp]: Late to the party, but anyway:

Either an enum, a local constant (`const bool allow_bucket = false;` and use that as argument) or a separate, aptly named function (that does not have the parameter and simply forwards to the actual function).

Anyway, code is fine as it is.

**Pull request merged**